File: assignment 1/Hybrid sort.py
import random
import time
arrays = {}
arrays[1000] = [random.randint(1, 1000) for _ in range(1000)]
arrays[5000] = [random.randint(1, 1000) for _ in range(5000)]

def merge(A, B):
    i = j = 0
    sorted_list = []
    while i < len(A) and j < len(B):
        if A[i] >= B[j]:
            sorted_list.append(A[i])
            i += 1
        else:
            sorted_list.append(B[j])
            j += 1
    return sorted_list + A[i:] + B[j:]


def insertion_sort(A):
    n = len(A)
    for i in range(1,n):
        for j in range(i,0,-1):
            if A[j] < A[j-1]:
                A[j], A[j-1] = A[j-1],A[j]
            else:
                break
    return A

def hybrid_sort(A,S):
    if len(A) <= 1:
        return A
    if len(A) <= S:
        return insertion_sort(A)
    mid = len(A)//2
    lower = A[:mid]
    higher = A[mid:]
    A1 = hybrid_sort(lower,S)
    A2 = hybrid_sort(higher,S)
    return merge(A1,A2)

for i in arrays.values():
    hybrid_sort(i,5)

File: assignment 1/java/HybridSort.java
public class HybridSort {
    private static long comparisons =0;

    public static void insertionSort(int[] arr, int left, int right) {
        for (int i = left + 1; i <= right; i++) {
            int key = arr[i];
            int j = i - 1;
            while (j >= left && arr[j] > key) {
            	comparisons++;
                arr[j + 1] = arr[j];
                j--;
            }
            arr[j + 1] = key;
        }
    }

    // Merge two subarrays
    public static void merge(int[] arr, int left, int mid, int right) {
        // Sizes of the two subarrays
        int n1 = mid - left + 1;
        int n2 = right - mid;

        // Create temporary arrays
        int[] leftArray = new int[n1];
        int[] rightArray = new int[n2];

        // Copy data to temp arrays
        for (int i = 0; i < n1; i++) {
            leftArray[i] = arr[left + i];
        }
        for (int j = 0; j < n2; j++) {
            rightArray[j] = arr[mid + 1 + j];
        }

        // Merge the temporary arrays
        int i = 0, j = 0;
        int k = left;

        while (i < n1 && j < n2) {
        	comparisons++;
            if (leftArray[i] <= rightArray[j]) {
                arr[k] = leftArray[i];
                i++;
            } else {
                arr[k] = rightArray[j];
                j++;
            }
            k++;
        }

        // Copy remaining elements of leftArray, if any
        while (i < n1) {
            arr[k] = leftArray[i];
            i++;
            k++;
        }

        // Copy remaining elements of rightArray, if any
        while (j < n2) {
            arr[k] = rightArray[j];
            j++;
            k++;
        }
    }

    public static void mergeSort(int[] arr, int left, int right) {
        if (left < right) {
            int mid = (left + right) / 2;

            // Sort first and second halves
            mergeSort(arr, left, mid);
            mergeSort(arr, mid + 1, right);

            // Merge the sorted halves
            merge(arr, left, mid, right);
        }
    }

    public static void hybridSort(int[] arr, int left, int right, int S) {
        if (right - left + 1 <= S) {
            insertionSort(arr, left, right);  // Switch to insertion sort for small subarrays
        } else {
            if (left < right) {
                int mid = (left + right) / 2;
                hybridSort(arr, left, mid, S);
                hybridSort(arr, mid + 1, right, S);
                merge(arr, left, mid, right);
            }
        }
    }

    public static long getComparisons() {
		return comparisons;
	}
	
	public static void setComparisons(long comp) {
		comparisons=comp;
	}
}
